src/main/scala/sangria/marshalling/ArrayMapBuilder.scala

```diff
@@ -62,4 +63,40 @@ class ArrayMapBuilder[T](keys: Seq[String]) {
 
     builder.result()
   }
+
+  lazy val toIterator: Iterator[(String, T)] = {
+    new Iterator[(String, T)] {
+      var index = -1
+      var nextIndex = 0
+
+      @tailrec def nextIndex(current: Int): Option[Int] = {
+        val next = current + 1
+        if (next >= elements.length) None
+        else if (indexesSet.contains(next)) Some(next)
+        else nextIndex(next)
+      }
+
+      override def hasNext: Boolean = {
+        nextIndex(index) match {
+          case Some(i) ⇒
+            nextIndex = i
+            true
+          case None ⇒
+            false
+        }
+      }
+
+      override def next(): (String, T) = {
+        index = nextIndex
```

Comment:
I'm not 100% sure about this, but i don't think that there is a contract that says something like: `hasNext` must be called before `next` is called. In the current state `next` will be stuck on the current element.

On the other hand, `next` scaladoc states:

> @return  the next element of this iterator, if `hasNext` is `true`, undefined behavior otherwise.

I guess it can be interpreted in favour of this approach.

Reply:
Good question.
I read somewhere that `hasNext()` is allowed to mutate the internal state. 
My assumption is that we **must** call `hasNext()` before calling `next()`.
If this assumption is wrong, then I'll have to re-thing the implementation.
I have to dig further here.

Reply:
> I read somewhere that hasNext() is allowed to mutate the internal state.

I can confirm this, it is described in the scaladoc (Iterator):

> Note that even `hasNext` may cause mutation

Reply:
After thinking about it, I think "undefined behavior otherwise" would cover this case. So if `hasNext` is not called, `next` is allowed to return the same element over and over again since behavior in this case is undefined.